### DISCUSSION TITLE
Faceted search on fanout index

### DISCUSSION THREAD

**Discussion by bruno-cerutti** [ravendb/ravendb#13292] [Q&A]
Hi, I have a class "Product" having many options with different attributes. For example the same product (a pair of shoes) can have 6 options with different attributes of Color and Size (Color: Redand Size: 9, Color: Red and Size: 10, Color: Black and Size: 9, etc.,)
You can find a working example here https://github.com/bruno-cerutti/FacetsTest
A simplified version can be this one:
```
public class Product
{
        public string Id { get; set; }
        public string Title { get; set; }
        public string Brand { get; set; }
        public IList<ProductOption> Options { get; set; }
    }
```
```
public class ProductOption
    {
        public int OptionId { get; set; }
        public bool Available { get; set; }
        public decimal Price { get; set; }
        public IList<ProductOptionAttribute> Attributes { get; set; }
    }
```
```
public class ProductOptionAttribute
    {
        public string Name { get; set; }
        public string Value { get; set; }
    }
```

I made an index to map all the properties I need to filter
```
public class Products_Options : AbstractIndexCreationTask<Product>
    {
        public Products_Options()
        {
            Map = products => from doc in products
                from option in doc.Options
                select new
                {
                    doc.Title,
                    doc.Brand,
                    Option_Available = option.Available,
                    _ = option.Attributes.Select(attribute => CreateField(attribute.Name, attribute.Value))
                };
        }
    }
```
Now I'd like to use facets to aggregate results by brand, availability, size and color but the result of every FacetResult is returning the count of the options that match and not the count based on the number of products.

What I was expecting is the following result:
1. When I'm using Product_Options grouping by brand I get a FacetResult with two FacetValue, each one with Count 6. What I was expecting was a FacetResult with two FacetValue, each one with Count 1.
2. When I'm using Product_Options filtering by availability=true and Size=9, grouping by brand I get a FacetResult with one FacetValue with Count 2. What I was expecting was a FacetResult with one FacetValue with Count 1.

Is there any chance to achieve what I was expecting through index and not via in memory implementation?

Thanks in advance!

**Comment by ayende** [comment]
You are actually indexing the options, not the products, so this is why we are getting this situation. 
Instead, select this once, and return an array for the whole options, like so:

```
public class Products_Options : AbstractIndexCreationTask<Product>
{
        public Products_Options()
        {
            Map = products => from doc in products
                select new
                {
                    doc.Title,
                    doc.Brand,
                    _ = doc.Options.Select(option => option.Attributes.Select(attribute => CreateField(attribute.Name, attribute.Value)))
                };
        }
}
```

This will give you a single document for each product, and facets should work properly.

**Comment by DejanMilicic** [comment]
@bruno-cerutti I modified index, please look at my fork https://github.com/DejanMilicic/FacetsTest

**Comment by bruno-cerutti** [comment]
Thank you for your replies!
I tried to use the index modified but now the filtering doesn't work as I want.
If I filter by Size 9 and Availability true, as I do in my example, I expect to have as result only the shoes that has an option available with size 9. With the modified index I have as result both the shoes, despite one of these doesn't have any options available with size 9.

**Reply by ayende** [reply]
Can you show the data and the query you are using? I don't follow.
A test case would be great.

**Reply by bruno-cerutti** [reply]
I updated my example https://github.com/bruno-cerutti/FacetsTest with some tests where I describe the expected result.
I added in that project the index modified as you (and @DejanMilicic) have suggested with the name of "Products_Options_V2".
In ProductsQueries.cs there's a method "InitializeCollection" that saves to Raven two sample products.

**Comment by DejanMilicic** [comment]
I think the tests are not correct. Here is a short summary of products and tests as they should be

nike, false, 9, red
nike, true, 10, red
nike, false, 11, red
nike, false, 9, black
nike, true, 10, black
nike, false, 11, black
adidas, true, 9, red
adidas, false, 11, red
adidas, true, 9, yellow
adidas, true, 10, yellow
adidas, false, 11, yellow
adidas, true, 12, yellow 

1st test: "unavailable"
"unavailable" brands: 2
"unavailable" adidas products: 2 (11-red, 11-yellow)
"unavailable" nike products: 4 (9-red, 11-red, 9-black, 11-black)

2nd test: "available, size 9"
"available, size 9" brands: 1 (adidas)
"available, size 9" products: 2 (adidas 9-red, adidas 9-yellow)

3rd test: "available, size 11"
"available, size 11" brands: 0

I updated index and tests at https://github.com/DejanMilicic/FacetsTest

**Reply by DejanMilicic** [reply]
@bruno-cerutti did you have time to check?
